src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -3453,6 +3453,26 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
     }
 #endif // PROFILING_SUPPORTED
 
+    // If m_pThreadAttemptingSuspendForGC is set, it means:
+    // 1) someone is trying to suspend for GC and
+    // 2) it is not us.
+    // In such case, before doing much work, we back off with ERROR_TIMEOUT and will try again later
+    // This should be rare, but if too many non-GC suspensions are trying to happen, GC could starve.
+    if (m_pThreadAttemptingSuspendForGC != NULL)
+    {
+#ifdef PROFILING_SUPPORTED
+        // Must let the profiler know that this thread is aborting its attempt at suspending
+        {
+            BEGIN_PIN_PROFILER(CORProfilerTrackSuspends());
+            g_profControlBlock.pProfInterface->RuntimeSuspendAborted();
```

Comment:
With this new location of the suspend abort, it doesn't appear that it protects against a deadlock on the threadstore lock. Could you explain the reasoning here?

Reply:
I do not think suspension-yielding behavior was done to guarantee that GC suspension will preempt other suspension, because it did not do that. 
The thread that requests GC suspension is always in preemptive mode and would be ignored by the thread doing suspension. GC suspender can signal about its intentions by setting a global flag, but that could easily be too late for the suspending thread to notice.

I believe the purpose is a mitigation for situations where we have a _lot_ of non-GC suspensions. I am not sure how common/real such scenarios are, but it could potentially starve GC so we have this suspension-yielding behavior. 
The other part of this mechanism is an event that non-GC suspenders could be forced to wait on while GC suspension is in progress. That is also racey - when we reset the event we may already have non-GC suspenders waiting on the TS lock and not affected by the event.

Basically - this mechanism clearly does not guarantee that _any_ non-GC suspension will immediately yield to GC suspension. At best it can guarantee that if we have 100 non-GC suspenders, then a GC suspension will gain higher priority and get ahead most of them.
For such purpose checking once before starting the suspension loop is just as good as checking at each iteration.

I.E. we may allow one non-GC suspension that is already half-way or almost done, but will deprioritize others and this should still serve the starvation-prevention purpose.

Doing the check at this point has a number of benefits though:
- suspending EE is a fairly heavy operation. It makes sense to not abort suspension that is nearly done. (and we can't abort one that is completely done anyways).
- it minimizes the number of states we have to deal with when restarting EE after an aborted suspension - we would not have, for example threads that are already redirected oh hijacked and asynchronously moving towards suspension as we decide to resume or suspend again.

Reply:
The removed comment from the previous implementation discusses a deadlock, where the GC thread attempts to get the threadstore lock, but a separate suspended thread already holds the lock. Effectively, I think I'm asking, can that deadlock occur?

Reply:
It looks like this code had a complicated history and some comments no longer match what code does.
I assume the question is about:

```c
            // When the debugger is synchronizing, trying to perform a GC could deadlock. The GC has the
            // threadstore lock and synchronization cannot complete until the debugger can get the
            // threadstore lock. However the GC can not complete until it sends the BeforeGarbageCollection
            // event, and the event can not be sent until the debugger is synchronized. In order to break
            // this deadlock cycle the GC must give up the threadstore lock, allow the debugger to synchronize,
            // then try again.
```

I do not think the scenario described in the comment exists.

GC does not post `BeforeGarbageCollection` event. At least I could not find when it does. Perhaps it did in the past.

Regardless, debugger should not be synchronizing when GC holds the lock.
I have added an assert, just in case, but I think it is not possible for a GC suspension to be happening while debugger is `IsSynchronizing`. 

Right now switching into `IsSynchronizing` mode happens when debugger already owns the TSL. here: https://github.com/dotnet/runtime/blob/4486164506a56200e66cf4cd750fde6b54ab5abd/src/coreclr/src/debug/ee/debugger.cpp#L5503 
It is done via an ordinary write but it is done while holding TSL so that is ok. All places that handle this flag appear to hold the TSL as well, except when it is reset - not sure if we hold the TSL at that point, but contracts do not require that.

There is only one case when GC could yield the suspension - when debugger has threads in inconsistent state - like stopped on a breakpoint. Then the thread may appear stopped on a safe point, but actually it is not. 
This is the only case where everyone, including GC must retry suspending - since suspension did not really succeed.
We still handle this scenario. 
The state is indicated by `g_pDebugInterface->ThreadsAtUnsafePlaces()`, which is an interlock-updated counter. We read the value after EE is suspended and thus we read a stable value.

Reply:
>  Regardless, debugger should not be synchronizing when GC holds the lock.

If above is actually not true, we should check for that _before_ doing suspension - same place where we check for GC vs. non-GC suspension collisions. At that point we hold TSL, so the value will not change, but we have not done much work yet.
I do not think it is a case that we need to handle though.

Reply:
There was another comment:

```
        // must resume all of the threads and tell the GC that we are
        // at a safepoint - since this is the exact same behaviour
        // that the debugger needs, just use it's code.
```

I am not sure what that refers to. What it suggests seems unsafe right now. 
We can only assume a safe point if suspension succeeds, thus anyone trying to suspend will either succeed or retry in a loop. Perhaps it was different in the past.

Reply:
Actually GC does send `BeforeGarbageCollection`. Just found the place. 
I am still not sure if we can find debugger in `IsSynchronizing` state at that point. Let me check more on that.

Reply:
Ah, sorry, no, this was in reference to... 
```
        // If another thread is trying to do a GC, there is a chance of deadlock
        // because this thread holds the threadstore lock and the GC thread is stuck
        // trying to get it, so this thread must bail and do a retry after the GC completes.
        //
```

I took a quick look at the GC suspend logic, and it appears that it sets m_pThreadAttemptingSuspendForGC and then takes the threadstore lock on the GC thread. While with your current change, the thread suspending, takes the lock, then checks m_pThreadAttemptingSuspendForGC. This can lead to the currently suspending thread checking m_pThreadAttemptingSuspendForGC, and then, on the GC thread, setting that value, then waiting on the ThreadStore lock. That's the deadlock I'm concerned about.

Reply:
> This can lead to the currently suspending thread checking m_pThreadAttemptingSuspendForGC, and then, on the GC thread, setting that value

I think that was possible before too. 
The thread that sets `m_pThreadAttemptingSuspendForGC` is in preemptive mode and can be arbitrary late related to the other thread reading that value.

If it was coop, then we could assume that the other thread would keep spinning until it sees the flag, but a thread suspending for GC is always preemptive.

Reply:
Imagine a context switch delaying the GC thread by XX time right before setting `m_pThreadAttemptingSuspendForGC`. 
Since GC thread is preemptive, the other thread could start, continue, finish the suspension, or even finish that what it was going to do and resume EE, if XX is long enough.

That is the case before or after the change.

Reply:
The issue is that the thread that sets m_pThreadAttemptingSuspendForGC is in cooperative mode not preemptive mode. When a GC is triggering a suspend, it is in cooperative mode (as its in the process of manipulating GC data structures), and the LockThreadStore routine does not transition to preemptive mode when attempting to suspend for a GC.  (See the toggleGC boolean in ThreadSuspend::LockThreadStore.) Thus there is a potential for deadlock.

The old logic had this abort routine, which would kick in at any point in time during suspend, and if it ever triggered, it would stop attempting to suspend. Effectively, this make the whole threadstore a deadlock detecting lock, with a policy favoring the GC thread suspension logic. There was a race here, but it is protected by the abort process being inside the loop, and being able to abort suspension whenever the thread attempting suspend status became true. Effectively, the check could race, but only for one extra iteration of the suspension loop.

Now, the previous logic would simply abort the suspend immediately, and leave the various threads that it had managed in a state where they were suspended for some other purpose, and then `ThreadSuspend::SuspendEE` will call `RestartEE` and then use a goto to exit the threadlock, and then hopefully the the GC thread would come along and finish the suspend from its end. That logic all seems a bit sketchy to me, but it looks like it would have worked in the old approach. With your new code that moved the abort logic out of the loop, I don't see how this works.

Reply:
I think LockThreadStore does not transition to preemptive for GC threads because they are already supposed to be preemptive. Otherwise we could block on TSL while in coop mode, which would deadlock. 

Let me check. It may be worth adding asserts/contracts to be explicit about this.

Reply:
Actually I have added a contract to `ThreadStore::Enter()` for exactly this purpose already, I just forgot. 

Probably worth adding a comment to LockThreadStore on why we do not bother about GC threads. 
Or better not specialcase them in the toggleGC check at all.

Reply:
We discussed this offline, and I'm convinced that the gc thread is preemptive when suspending.